router/pkg/connectrpc/vanguard_service.go

```diff
@@ -0,0 +1,374 @@
+
+		// Use NewServiceWithSchema with custom type resolver
+		// This avoids relying on the global registry
+		servicePath := "/" + serviceName + "/"
+
+		vs.logger.Debug("creating service with custom type resolver",
+			zap.String("service_path", servicePath))
+
+		// Configure to always transcode to Connect protocol with JSON codec
+		// This ensures our handler always receives JSON, regardless of the incoming protocol
+		// Use NewServiceWithSchema to provide the schema directly with a custom type resolver
+		vanguardService := vanguard.NewServiceWithSchema(
+			serviceDef.ServiceDescriptor,
+			serviceHandler,
+			vanguard.WithTargetProtocols(vanguard.ProtocolConnect),
+			vanguard.WithTargetCodecs("json"),
+			vanguard.WithTypeResolver(customTypes),
+		)
+
+		vs.services = append(vs.services, vanguardService)
+
+		vs.logger.Debug("registered service successfully with custom type resolver",
+			zap.String("service", serviceName),
+			zap.String("service_path", servicePath),
+			zap.String("target_protocol", "connect"),
+			zap.String("target_codec", "json"))
+	}
+
+	return nil
+}
+
+// createServiceHandler creates an HTTP handler for a specific proto service
+// This handler is wrapped by Vanguard, which handles protocol transcoding
+func (vs *VanguardService) createServiceHandler(serviceName string, serviceDef *ServiceDefinition) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Extract and validate method name from path
+		methodName := vs.extractMethodName(r.URL.Path, serviceName)
+		if methodName == "" {
+			// Return Connect error for invalid path
+			connectErr := connect.NewError(connect.CodeNotFound, fmt.Errorf("invalid path format"))
+			vs.writeConnectError(w, connectErr, serviceName, methodName)
+			return
+		}
+
+		// Validate method exists in service
+		methodExists := false
+		for _, method := range serviceDef.Methods {
+			if method.Name == methodName {
+				methodExists = true
+				break
+			}
+		}
+
+		if !methodExists {
+			// Return Connect error for method not found
+			connectErr := connect.NewError(connect.CodeNotFound, fmt.Errorf("method not found: %s", methodName))
+			vs.writeConnectError(w, connectErr, serviceName, methodName)
+			return
+		}
+
+		// For GET requests (Connect protocol), extract message from query parameter
+		// For POST requests, read from body
+		var requestBody []byte
+		var err error
+
+		if r.Method == "GET" {
+			// Extract the 'message' query parameter (Connect protocol for GET requests)
+			messageParam := r.URL.Query().Get("message")
+			if messageParam == "" {
+				// For methods with no input parameters, use empty JSON object
+				requestBody = []byte("{}")
+			} else {
+				// The message parameter is already URL-decoded by r.URL.Query().Get()
+				requestBody = []byte(messageParam)
+			}
+			vs.logger.Debug("extracted message from GET query parameter",
+				zap.String("message", string(requestBody)))
+		} else {
```

Comment:
Is it always POST? could it be OPTIONS? no clue, just asking

Reply:
Good point. We don't yet support CORS - I'll add it.